File: tools/cvf-skills-cli/src/bm25.js
/**
 * BM25 Scoring Engine — pure Node.js, zero dependencies.
 */

export class BM25 {
  constructor(k1 = 1.5, b = 0.75) {
    this.k1 = k1;
    this.b = b;
    this.corpus = [];
    this.docCount = 0;
    this.avgLengths = {};
    this.idfCache = {};
    this.fieldWeights = {};
  }

  index(documents, fields, weights = null) {
    this.fieldWeights = weights || Object.fromEntries(fields.map(f => [f, 1.0]));
    this.corpus = [];

    for (const doc of documents) {
      const tokenized = {};
      for (const field of fields) {
        tokenized[field] = this._tokenize(doc[field] || '');
      }
      this.corpus.push(tokenized);
    }

    this.docCount = this.corpus.length;

    for (const field of fields) {
      const lengths = this.corpus.map(d => (d[field] || []).length);
      this.avgLengths[field] = lengths.reduce((a, b) => a + b, 0) / Math.max(lengths.length, 1);
    }

    this._computeIdf(fields);
  }

  search(query, topN = 10) {
    const queryTokens = this._tokenize(query);
    if (queryTokens.length === 0) return [];

    const scores = [];

    for (let i = 0; i < this.corpus.length; i++) {
      const doc = this.corpus[i];
      let score = 0;

      for (const [field, tokens] of Object.entries(doc)) {
        const weight = this.fieldWeights[field] || 1.0;
        score += this._scoreField(queryTokens, tokens, field) * weight;
      }

      if (score > 0) scores.push([i, score]);
    }

    scores.sort((a, b) => b[1] - a[1]);
    return scores.slice(0, topN);
  }

  _tokenize(text) {
    if (!text) return [];
    const matches = text.toLowerCase().match(/[a-zA-ZÀ-ỹ0-9]+/g);
    return matches ? matches.filter(t => t.length >= 2) : [];
  }

  _computeIdf(fields) {
    const termDocCount = {};

    for (const doc of this.corpus) {
      const docTerms = new Set();
      for (const field of fields) {
        for (const token of (doc[field] || [])) docTerms.add(token);
      }
      for (const term of docTerms) {
        termDocCount[term] = (termDocCount[term] || 0) + 1;
      }
    }

    for (const [term, df] of Object.entries(termDocCount)) {
      this.idfCache[term] = Math.log((this.docCount - df + 0.5) / (df + 0.5) + 1.0);
    }
  }

  _scoreField(queryTokens, fieldTokens, field) {
    if (fieldTokens.length === 0) return 0;

    const docLen = fieldTokens.length;
    const avgLen = this.avgLengths[field] || 1;

    const tfMap = {};
    for (const token of fieldTokens) {
      tfMap[token] = (tfMap[token] || 0) + 1;
    }

    let score = 0;
    for (const qt of queryTokens) {
      const idf = this.idfCache[qt];
      if (idf === undefined) continue;
      const tf = tfMap[qt] || 0;
      if (tf === 0) continue;

      const numerator = tf * (this.k1 + 1);
      const denominator = tf + this.k1 * (1 - this.b + this.b * docLen / Math.max(avgLen, 1));
      score += idf * numerator / denominator;
    }

    return score;
  }
}

File: tools/skill-search/bm25.py
#!/usr/bin/env python
"""BM25 scoring engine for CVF Skill Library search.

Pure Python implementation — no external dependencies.
"""
from __future__ import annotations

import math
from typing import Dict, List, Tuple


class BM25:
    """Okapi BM25 ranking algorithm.

    Parameters:
        k1: Term frequency saturation (default 1.5)
        b:  Document length normalization (default 0.75)
    """

    def __init__(self, k1: float = 1.5, b: float = 0.75):
        self.k1 = k1
        self.b = b
        self.corpus: List[Dict[str, List[str]]] = []  # [{field: [tokens]}]
        self.doc_count = 0
        self.avg_lengths: Dict[str, float] = {}
        self.idf_cache: Dict[str, float] = {}
        self.field_weights: Dict[str, float] = {}

    def index(
        self,
        documents: List[Dict[str, str]],
        fields: List[str],
        weights: Dict[str, float] | None = None,
    ) -> None:
        """Build index from list of document dicts.

        Args:
            documents: List of {field_name: text_value}
            fields: Fields to index
            weights: Field name → weight multiplier
        """
        self.field_weights = weights or {f: 1.0 for f in fields}
        self.corpus = []

        for doc in documents:
            tokenized = {}
            for field in fields:
                text = doc.get(field, "")
                tokenized[field] = self._tokenize(text)
            self.corpus.append(tokenized)

        self.doc_count = len(self.corpus)

        # Average field lengths
        for field in fields:
            lengths = [len(d.get(field, [])) for d in self.corpus]
            self.avg_lengths[field] = sum(lengths) / max(len(lengths), 1)

        # Pre-compute IDF for all terms
        self._compute_idf(fields)

    def search(
        self,
        query: str,
        top_n: int = 10,
    ) -> List[Tuple[int, float]]:
        """Search the indexed corpus.

        Args:
            query: Search query string
            top_n: Maximum results to return

        Returns:
            List of (doc_index, score) tuples, sorted by score descending
        """
        query_tokens = self._tokenize(query)
        if not query_tokens:
            return []

        scores: List[Tuple[int, float]] = []

        for doc_idx, doc in enumerate(self.corpus):
            score = 0.0
            for field, tokens in doc.items():
                weight = self.field_weights.get(field, 1.0)
                field_score = self._score_field(query_tokens, tokens, field)
                score += field_score * weight
            if score > 0:
                scores.append((doc_idx, score))

        scores.sort(key=lambda x: x[1], reverse=True)
        return scores[:top_n]

    def _tokenize(self, text: str) -> List[str]:
        """Tokenize text into lowercase words."""
        if not text:
            return []
        # Split on non-alphanumeric (keep Vietnamese diacritics)
        import re
        tokens = re.findall(r"[a-zA-ZÀ-ỹ0-9]+", text.lower())
        return [t for t in tokens if len(t) >= 2]

    def _compute_idf(self, fields: List[str]) -> None:
        """Pre-compute inverse document frequency for all terms."""
        term_doc_count: Dict[str, int] = {}

        for doc in self.corpus:
            # Collect unique terms across all fields in this doc
            doc_terms: set = set()
            for field in fields:
                doc_terms.update(doc.get(field, []))

            for term in doc_terms:
                term_doc_count[term] = term_doc_count.get(term, 0) + 1

        for term, df in term_doc_count.items():
            # Standard BM25 IDF with smoothing
            self.idf_cache[term] = math.log(
                (self.doc_count - df + 0.5) / (df + 0.5) + 1.0
            )

    def _score_field(
        self,
        query_tokens: List[str],
        field_tokens: List[str],
        field: str,
    ) -> float:
        """Score a single field against query tokens."""
        if not field_tokens:
            return 0.0

        doc_len = len(field_tokens)
        avg_len = self.avg_lengths.get(field, 1.0)

        # Term frequency in this field
        tf_map: Dict[str, int] = {}
        for token in field_tokens:
            tf_map[token] = tf_map.get(token, 0) + 1

        score = 0.0
        for qt in query_tokens:
            if qt not in self.idf_cache:
                continue
            idf = self.idf_cache[qt]
            tf = tf_map.get(qt, 0)
            if tf == 0:
                continue

            # BM25 formula
            numerator = tf * (self.k1 + 1)
            denominator = tf + self.k1 * (1 - self.b + self.b * doc_len / max(avg_len, 1))
            score += idf * numerator / denominator

        return score
